File: internal/aistudio/value.go
package aistudio

import "fmt"

func EncodeStruct(value any) any {
	object, ok := value.(map[string]any)
	if !ok {
		object = map[string]any{}
	}
	entries := make([]any, 0, len(object))
	for key, item := range object {
		entries = append(entries, []any{key, EncodeValue(item)})
	}
	return []any{entries}
}
func EncodeValue(value any) any {
	switch item := value.(type) {
	case nil:
		return []any{0}
	case bool:
		return []any{nil, nil, nil, item}
	case float64:
		return []any{nil, item}
	case int:
		return []any{nil, item}
	case string:
		return []any{nil, nil, item}
	case map[string]any:
		return []any{nil, nil, nil, nil, EncodeStruct(item)}
	case []any:
		values := make([]any, 0, len(item))
		for _, nested := range item {
			values = append(values, EncodeValue(nested))
		}
		return []any{nil, nil, nil, nil, nil, []any{values}}
	default:
		panic(fmt.Sprintf("Unsupported protobuf Struct value: %T", value))
	}
}

File: src/aistudio_client/makersuite/value.py
"""مقدار JSON را به google.protobuf.Value positional تبدیل می‌کند."""

from typing import Any


def encode_struct(value: dict[str, Any]) -> list:
    return [[[key, encode_value(item)] for key, item in value.items()]]


def encode_value(value: Any) -> list:
    if value is None:
        return [0]
    if isinstance(value, bool):
        return [None, None, None, value]
    if isinstance(value, (int, float)):
        return [None, value]
    if isinstance(value, str):
        return [None, None, value]
    if isinstance(value, dict):
        return [None, None, None, None, encode_struct(value)]
    if isinstance(value, list):
        return [None, None, None, None, None, [[encode_value(item) for item in value]]]
    raise ValueError(f"Unsupported protobuf Struct value: {type(value).__name__}")
